File: packages/engine/src/core/told.ts
import { mkdirSync, readFileSync, rmSync } from 'node:fs'
import { join } from 'node:path'
import { writeFileAtomic } from './atomic'
import { ensureIndexDir, indexDir } from './index-store'

/**
 * What a session has already been told (memory-lead 2.1, D6): one entry per
 * (decision, subject) that a PostToolUse notice named, on a read or an edit.
 *
 * Reads append nothing to the record, so unlike the edit's lastTouch there is
 * no event to reconstruct this from. It lives in the derived index instead,
 * which makes it disposable in the safe direction: a lost, corrupt or raced
 * file re-tells a decision and never silences one. SessionStart deletes the
 * file on `compact` and `clear`, because the context that held the notices is
 * gone.
 */

const TOLD_DIR = 'told'
const TOLD_VERSION = 1

/** One told pair. The subject is the repo-relative path the notice named. */
export function toldKey(decisionId: string, subject: string): string {
  return `${decisionId} ${subject}`
}

function toldFile(sofarDir: string, session: string): string {
  return join(indexDir(sofarDir), TOLD_DIR, `${session.replace(/[^A-Za-z0-9_-]/g, '_')}.json`)
}

/** The pairs this session was told; empty for `cli` or when nothing usable is on disk. */
export function readTold(sofarDir: string, session: string): Set<string> {
  if (session === 'cli') return new Set()
  try {
    const raw = JSON.parse(readFileSync(toldFile(sofarDir, session), 'utf8')) as { v?: unknown; told?: unknown }
    if (raw.v !== TOLD_VERSION || !Array.isArray(raw.told)) return new Set()
    return new Set(raw.told.filter((k): k is string => typeof k === 'string'))
  } catch {
    return new Set()
  }
}

/** Add pairs to the session's set. Silent on failure: the cost of a lost write is one repeat. */
export function addTold(sofarDir: string, session: string, keys: readonly string[]): void {
  if (session === 'cli' || keys.length === 0) return
  try {
    const told = readTold(sofarDir, session)
    for (const key of keys) told.add(key)
    ensureIndexDir(sofarDir)
    mkdirSync(join(indexDir(sofarDir), TOLD_DIR), { recursive: true })
    writeFileAtomic(toldFile(sofarDir, session), `${JSON.stringify({ v: TOLD_VERSION, told: [...told] })}\n`)
  } catch {
    // See the header: a set that cannot be written re-tells.
  }
}

/** Forget what the session was told, because its context was compacted or cleared. */
export function clearTold(sofarDir: string, session: string): void {
  try {
    rmSync(toldFile(sofarDir, session), { force: true })
  } catch {
    // Nothing to forget.
  }
}

File: crates/sofar-core/src/told.rs
//! What a session has already been told (`core/told.ts`, memory-lead 2.1,
//! D6): one entry per (decision, subject) a `PostToolUse` notice named, on a
//! read or an edit. Reads append nothing to the record, so this lives in the
//! derived index, disposable in the safe direction: a lost, corrupt or raced
//! file re-tells a decision and never silences one. `SessionStart` deletes it
//! on `compact` and `clear`, because the context that held the notices is
//! gone.

use std::fs;
use std::path::PathBuf;

use crate::atomic::write_file_atomic;
use crate::json::{self, Json, Object};
use crate::layout::Layout;

const TOLD_DIR: &str = "told";
const TOLD_VERSION: f64 = 1.0;

/// `toldKey`: one told pair; the subject is the repo-relative path the notice named.
#[must_use]
pub fn told_key(decision_id: &str, subject: &str) -> String {
    format!("{decision_id} {subject}")
}

/// `session.replace(/[^A-Za-z0-9_-]/g, '_')`, per UTF-16 unit.
fn safe_session(session: &str) -> String {
    session
        .encode_utf16()
        .map(|u| match char::from_u32(u32::from(u)) {
            Some(c) if c.is_ascii_alphanumeric() || c == '_' || c == '-' => c,
            _ => '_',
        })
        .collect()
}

fn told_file(layout: &Layout, session: &str) -> PathBuf {
    layout
        .index_dir()
        .join(TOLD_DIR)
        .join(format!("{}.json", safe_session(session)))
}

/// `readTold`: the pairs this session was told, in insertion order; empty for
/// `cli` or when nothing usable is on disk.
#[must_use]
pub fn read_told(layout: &Layout, session: &str) -> Vec<String> {
    if session == "cli" {
        return Vec::new();
    }
    let Ok(bytes) = fs::read(told_file(layout, session)) else {
        return Vec::new();
    };
    let Ok(Json::Obj(raw)) = json::parse(&String::from_utf8_lossy(&bytes)) else {
        return Vec::new();
    };
    if raw.get("v") != Some(&Json::Num(TOLD_VERSION)) {
        return Vec::new();
    }
    let Some(told) = raw.get("told").and_then(Json::as_arr) else {
        return Vec::new();
    };
    let mut out: Vec<String> = Vec::new();
    for key in told.iter().filter_map(Json::as_str) {
        if !out.iter().any(|k| k == key) {
            out.push(key.to_owned());
        }
    }
    out
}

/// `addTold`: add pairs to the session's set. Silent on failure: the cost of
/// a lost write is one repeat.
pub fn add_told(layout: &Layout, session: &str, keys: &[String]) {
    if session == "cli" || keys.is_empty() {
        return;
    }
    let mut told = read_told(layout, session);
    let mut seen: std::collections::HashSet<String> = told.iter().cloned().collect();
    for key in keys {
        if seen.insert(key.clone()) {
            told.push(key.clone());
        }
    }
    let Ok(dir) = layout.ensure_index_dir() else {
        return;
    };
    if fs::create_dir_all(dir.join(TOLD_DIR)).is_err() {
        return;
    }
    let mut o = Object::with_capacity(2);
    o.insert("v", Json::Num(TOLD_VERSION));
    o.insert("told", Json::Arr(told.into_iter().map(Json::Str).collect()));
    let mut text = json::stringify(&Json::Obj(o));
    text.push('\n');
    let _ = write_file_atomic(&told_file(layout, session), text.as_bytes());
}

/// `clearTold`: forget what the session was told (its context was compacted or cleared).
pub fn clear_told(layout: &Layout, session: &str) {
    let _ = fs::remove_file(told_file(layout, session));
}
